thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/task/TaskDriver.java

```diff
@@ -98,21 +105,33 @@ public void start() throws Exception {
               ThirdeyeMetricsUtil.taskCounter.inc();
 
               try {
-                LOG.info("Executing task: {} {}", anomalyTaskSpec.getJobName(),
-                    anomalyTaskSpec.getTaskInfo());
+                LOG.info("Executing task: {} {}", anomalyTaskSpec.getJobName(), anomalyTaskSpec.getTaskInfo());
 
                 // execute the selected task
                 TaskType taskType = anomalyTaskSpec.getTaskType();
                 TaskRunner taskRunner = TaskRunnerFactory.getTaskRunnerFromTaskType(taskType);
                 TaskInfo taskInfo = TaskInfoFactory.getTaskInfoFromTaskType(taskType, anomalyTaskSpec.getTaskInfo());
-
                 updateTaskStartTime(anomalyTaskSpec.getId());
-                List<TaskResult> taskResults = taskRunner.execute(taskInfo, taskContext);
+                Future<List<TaskResult>> future = taskExecutorService.submit(new Callable<List<TaskResult>>() {
+                  @Override
+                  public List<TaskResult> call() throws Exception {
+                    return taskRunner.execute(taskInfo, taskContext);
+                  }
+                });
+                try {
+                  List<TaskResult> taskResults = future.get(driverConfiguration.getMaxTaskRunTimeMillis(), TimeUnit.MILLISECONDS);
+                } catch (TimeoutException e) {
```

Comment:
Should we merge the try-catch with the outter level try-catch block? That looks cleaner.

Reply:
It cannot be put in the outside catch block, because we need to use the `future` instance to cancel the timeout thread, and it is not accessible there.